text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,475 @@
+
+### Changing from `panic=unwind` to `panic=abort` cannot cause UB
+
+This constraint is met:
+
+* Unwinding across a "C" boundary is UB regardless
+    of whether one is using `panic=unwind` or `panic=abort`.
+* Unwinding across a "C unwind" boundary is always defined,
+    though it is defined to abort if `panic=abort` is used.
+* Forced exceptions behave the same regardless of panic mode.
+
+### Optimization with panic=abort
+
+Using this proposal, the compiler is **almost always** able to reduce
+overhead related to unwinding when using panic=abort. The one
+exception is that invoking a "C unwind" ABI still requires some kind
+of minimal landing pad to trigger an abort. The expectation is that
+very few functions will use the "C unwind" boundary unless they truly
+intend to unwind -- and, in that case, those functions are likely
+using panic=unwind anyway, so this is not expected to make much
+difference in practice.
+
+### Preserve the ability to change how Rust panics are propagated when using the Rust ABI
+
+This constraint is met. If we were to change Rust panics to a
+different mechanism from the mechanism used by the native ABI,
+however, there would have to be a conversion step that interconverts
+between Rust panics and foreign exceptions at "C unwind" ABI
+boundaries.
+
+### Enable Rust panics to traverse through foreign frames
+
+This constraint is met.
+
+### Enable foreign exceptions to propagate through Rust frame
+
+This constraint is partially met: the behavior of foreign exceptions
+with respect to `catch_unwind` is currently undefined, and left for
+future work.
+
+### Enable error handling with `longjmp`
+
+This constraint has been [deferred][unresolved-questions].
+
+### Do not change the ABI of functions in the `libc` crate
+
+This constraint has been [deferred][unresolved-questions].
+
+# Prior art
+[prior-art]: #prior-art
+
+C++ as specified has no concept of "foreign" exceptions or of an underlying
+exception mechanism. However, in practice, the C++ exception mechanism is the
+"native" unwinding mechanism used by compilers.
+
+On Microsoft platforms, when using MSVC, unwinding is always supported for both
+C++ and C code; this is very similar to "option 3" described in [the
+inside-rust post][inside-rust-proposals] mentioned [above][alternatives].
+
+On other platforms, GCC, LLVM, and any related compilers provide a flag,
+`-fexceptions`, for explicitly ensuring that stack frames have unwinding
+support regardless of the language being compiled. Conversely,
+`-fno-exceptions` removes unwinding support even from C++. This is somewhat
+similar to how Rust's `panic=unwind` and `panic=abort` work for `panic!`
+unwinds, and under the "option 3" proposal, the behavior would be similar for
+foreign exceptions as well. In the current proposal, though, such foreign
+exception support is not enabled by default with `panic=unwind` but requires
+the new `"C unwind"` ABI.
+
+## Attributes on nightly Rust and prior RFCs
```

Comment:
As I mentioned on the previous RFC (https://github.com/rust-lang/rfcs/pull/2753/files#r325564278), replacing existing uses of `#[unwind]` will immediately require introducing `extern "stdcall unwind"` (aka `extern "system unwind"`) and `extern "thiscall unwind"` in addition to `extern "C unwind"`, because that's what libunwind uses (https://github.com/rust-lang/rust/blob/master/src/libpanic_unwind/seh.rs).

Reply:
Hm, I thought @Amanieu had suggested that this wouldn't be necessary for some reason, but I can't actually find that conversation.

I have a note from a March 3rd sync-up with the lang team saying that `extern "system unwind"` is "probably not worth supporting" since it's only used on x86 Windows. But clearly throwing SEH on x86 Windows through Rust frames is in fact required by the `panic` implementation!

Perhaps we should _only_ support `extern "system unwind"` for now, since it's equivalent to `"C unwind"` everywhere except x86 Windows?

What is `thiscall`? I don't see it in the list of supported ABI strings in the Nomicon.

Reply:
https://en.wikipedia.org/wiki/X86_calling_conventions#thiscall it is similar to stdcall, but has an extra `this` argument, which on x86 is passed in ECX.

Reply:
@petrochenkov I've added these three ABIs to the RFC.